ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -806,32 +854,98 @@ toCBORAHeaderOrBoundary pm epochSlots abob =
   Mempool integration
 -------------------------------------------------------------------------------}
 
+-- | The hash of a delegation certificate.
+type DlgCertId = Crypto.AbstractHash Blake2b_256 CC.Delegation.Certificate
+
+-- | The hash of an update vote.
+type UpdateVoteId = Crypto.AbstractHash Blake2b_256 CC.Update.Vote
+
+-- | An error type which represents either a UTxO, delegation, update proposal
+-- registration, or update vote error in the Byron era.
+data ByronApplyTxError
+  = ByronApplyTxError !CC.UTxO.UTxOValidationError
+  | ByronApplyDlgError !V.Scheduling.Error
+  | ByronApplyUpdateProposalError !CC.UPI.Error
+  | ByronApplyUpdateVoteError !CC.UPI.Error
+  deriving (Show)
+
+instance ToCBOR ByronApplyTxError where
+  toCBOR (ByronApplyTxError err) =
+    CBOR.encodeListLen 2 <> toCBOR (0 :: Word8) <> toCBOR err
+  toCBOR (ByronApplyDlgError err) =
+    CBOR.encodeListLen 2 <> toCBOR (1 :: Word8) <> toCBOR err
+  toCBOR (ByronApplyUpdateProposalError err) =
+    CBOR.encodeListLen 2 <> toCBOR (2 :: Word8) <> toCBOR err
+  toCBOR (ByronApplyUpdateVoteError err) =
+    CBOR.encodeListLen 2 <> toCBOR (3 :: Word8) <> toCBOR err
+
+instance FromCBOR ByronApplyTxError where
+  fromCBOR = do
+    enforceSize "ByronApplyTxError" 2
+    CBOR.decodeWord8 >>= \case
+      0   -> ByronApplyTxError             <$> fromCBOR
+      1   -> ByronApplyDlgError            <$> fromCBOR
+      2   -> ByronApplyUpdateProposalError <$> fromCBOR
+      3   -> ByronApplyUpdateVoteError     <$> fromCBOR
+      tag -> cborError $ DecoderErrorUnknownTag "ByronApplyTxError" tag
+
 instance (ByronGiven, Typeable cfg, ConfigContainsGenesis cfg)
       => ApplyTx (ByronBlockOrEBB cfg) where
   -- | Generalized transactions in Byron
   --
   -- TODO #514: This is still missing the other cases (this shouldn't be a
   -- newtype)
-  data GenTx (ByronBlockOrEBB cfg) = ByronTx
-    { byronTx   :: !(CC.UTxO.ATxAux ByteString)
-    , byronTxId :: CC.UTxO.TxId
-      -- ^ This field is lazy on purpose so that the TxId is computed on
-      -- demand.
-    }
-
-  newtype GenTxId (ByronBlockOrEBB cfg) = ByronTxId { unByronTxId :: CC.UTxO.TxId }
+  data GenTx (ByronBlockOrEBB cfg)
+    = ByronTx
+        CC.UTxO.TxId
+        -- ^ This field is lazy on purpose so that the 'CC.UTxO.TxId' is
+        -- computed on demand.
+        !(CC.UTxO.ATxAux ByteString)
+    | ByronDlg
+        DlgCertId
+        -- ^ This field is lazy on purpose so that the 'DlgCertId' is computed
+        -- on demand.
+        !(CC.Delegation.ACertificate ByteString)
+    | ByronUpdateProposal
+        CC.Update.UpId
+        -- ^ This field is lazy on purpose so that the 'CC.Update.UpId' is
+        -- computed on demand.
+        !(CC.Update.AProposal ByteString)
+    | ByronUpdateVote
+        UpdateVoteId
+        -- ^ This field is lazy on purpose so that the 'UpdateVoteId' is
+        -- computed on demand.
+        !(CC.Update.AVote ByteString)
+
+  data GenTxId (ByronBlockOrEBB cfg)
```

Comment:
Perhaps this should also be defined in `cardano-ledger` as `MempoolPayloadId` and then we could use a `newtype` here. This would save us from the longer definitions of `decodeByronGenTxId ` and `encodeByronGenTxId` as I'd define `ToCBOR` and `FromCBOR` instances for `MempoolPayloadId` as well.

Reply:
I see. Up to you.

Reply:
Perhaps it's fine to go with this for now and make changes later. However, it does seem quite appropriate to expose all of the required ID data types from `cardano-ledger`.